x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/async/AsyncTaskIndexService.java

```diff
@@ -284,25 +266,18 @@ private void updateResponse(
         ActionListener<UpdateResponse> listener,
         boolean isFailure
     ) {
+        ReleasableBytesStreamOutput buffer = null;
         try {
-            final ReleasableBytesStreamOutput buffer = isFailure
-                ? new ReleasableBytesStreamOutput(0, bigArrays.withCircuitBreaking())
-                : new ReleasableBytesStreamOutputWithLimit(0, bigArrays.withCircuitBreaking(), maxResponseSize);
-            final XContentBuilder source = XContentFactory.jsonBuilder(buffer);
-            listener = ActionListener.runBefore(listener, buffer::close);
-            source.startObject()
-                .field(RESPONSE_HEADERS_FIELD, responseHeaders)
-                .directFieldAsBase64(RESULT_FIELD, os -> writeResponse(response, os))
-                .endObject();
-            // do not close the buffer or the XContentBuilder until the UpdateRequest is completed (i.e., listener is notified);
-            // otherwise, we underestimate the memory usage in case the circuit breaker does not use the real memory usage.
-            source.flush();
-            final UpdateRequest request = new UpdateRequest().index(index)
-                .id(docId)
-                .doc(buffer.bytes(), source.contentType())
-                .retryOnConflict(5);
-            clientWithOrigin.update(request, listener);
+            buffer = allocateBuffer(isFailure == false);
+            final XContentBuilder source = XContentFactory.jsonBuilder(buffer).startObject().field(RESPONSE_HEADERS_FIELD, responseHeaders);
+            addResultFieldAndFinish(response, source);
+            clientWithOrigin.update(
+                new UpdateRequest().index(index).id(docId).doc(buffer.bytes(), source.contentType()).retryOnConflict(5),
+                ActionListener.runBefore(listener, buffer::close)
+            );
         } catch (Exception e) {
+            // release buffer right away to save memory, particularly in case the exception came from the circuit breaker
+            Releasables.close(buffer);
```

Comment:
This is the fix the PR description talks about. Without this change, we hold on to the buffer until the listener is finally resolved which could be a real problem because a `BigArrays` at capacity breaks all kinds of functionality.